### ISSUE TITLE
Pausing and continuation of experiments

### ISSUE THREAD

**Issue by tomkocmi** [ufal/neuralmonkey#201]
There should be a way how to end the training and later start at the point where the training was ended last time (the correct epoch, correct starting point in the corpus etc).

This way there should not be a difference when the model was trained as a whole and when it was paused during training (supposing the same random seed)

**Comment by jlibovicky** [comment]
I doesn't seem like an important feature to me. Can you give an example when it will be useful?
Would for instance allow skipping a given number of training instances do the job?

**Comment by martinpopel** [comment]
For example, we could do hyperparameter search with N epoch and M setups (where M is higher than what can be executed in parallel). After the N epochs, we would like to continue training the top K best-scoring setups for more epochs.
I think we can expect that training was not stopped in the middle of epoch, so _skipping a given number of training instances_ is not needed. We just need to load the model from checkpoint, including (decayed) learning rates etc.

**Comment by tomkocmi** [comment]
For example let say I want to train a model for a month, but during that time someone else will need the GPU for shorter period of time. Therefore I would stop my experiment and put it back to queue to give others the resources (GPU). I agree that skipping a given number of training instances would be only bonus and skipping only epochs would do the job.

**Comment by jindrahelcl** [comment]
So you are suggesting to implement a feature whose only job would be to print the correct epoch number to the log, when the training is resumed.

**Comment by martinpopel** [comment]
I would say "wrong" epoch number in the log is the last detail. @jindrahelcl: Have you checked that resumed training leads to exactly the same model as training in one run? If yes, then the only job is to document that NeMo already has this feature (and how to load existing models for training).
Note that e.g. in Vowpal Wabbit resumed learning leads to different model (because of learning rates, adaptive AdaGrad/Adam/whatever etc.) unless you specify `--save_resume`, which makes sure also this info is saved to the model.

**Comment by tomkocmi** [comment]
As I understand the learning_utils.py on line 107, they always beggin the experiment from epoch 0 no mather what you store anyhere. Am I incorrect?

**Comment by jindrahelcl** [comment]
But that's just the epoch **number**, it does not do anything else, it just loggs itself... Could you clarify what's the problem here, if it's not just the epoch number? Anyway, as we agreed a time ago, _continuing_ an experiment is an experiment itself and therefore the epoch numbering should start at zero.

I don't think this is an issue, nor believe the numbering should be changed.

**Comment by tomkocmi** [comment]
Depends on the reasons why did you stopped the experiment at the first place ... if you stop training to create a some checkpoint from which you want to train several times different experiments, then I agree with you ... but the reason why I want to pause it is to only allow the GPU for other computing and not to changing any setup, therefore I need a functionallity which will create identical results as in case that the experiment was stopped or run at the whole time.

**Comment by jindrahelcl** [comment]
I think this deserves to be closed. Feel free to reopen or create another issue, if there is one.

**Comment by loveJasmine** [comment]
Suppose I had training with epoch=10 with command 'neuralmonkey-train',  when it don, I found the model is not good enough, so I want continue to training more epoch, should I just start the new training with  command 'neuralmonkey-train' by the same ini again?

**Comment by jindrahelcl** [comment]
You need to edit the `[main]` section of the INI file,  set the `output` to
a different directory and add `initial_variables=/path/to/variables.data`
where the path should point to the original output directory. Note there is
actually no file called `variables.data`, but three files with this common
prefix. The initial_variables config value should correspond to this prefix.

2017-06-22 14:44 GMT+02:00 yk_data <notifications@github.com>:
